test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py

```diff
@@ -159,365 +119,173 @@ def get_field_types(cls):
-        return _is_bytes_n_instance_of(self, sub)
+    def __repr__(self):
+        return repr(self.items)
 
-    def __instancecheck__(self, other):
-        return _is_bytes_n_instance_of(self, other.__class__)
+    def __iter__(self):
+        return iter(self.items)
 
     def __eq__(self, other):
-        return _is_equal_bytes_n_type(self, other)
-
-    def __ne__(self, other):
-        return not _is_equal_bytes_n_type(self, other)
+        return self.items == other.items
 
-    def __hash__(self):
-        return hash(self.__class__)
-
-
-def parse_bytes(val):
-    if val is None:
-        return None
-    elif isinstance(val, str):
-        # TODO: import from eth-utils instead, and do: hexstr_if_str(to_bytes, val)
-        return None
-    elif isinstance(val, bytes):
-        return val
-    elif isinstance(val, int):
-        return bytes([val])
-    else:
-        return None
-
-
-class BytesN(bytes, metaclass=BytesNMeta):
-    def __new__(cls, *args):
-        if not hasattr(cls, 'length'):
-            return
-        bytesval = None
-        if len(args) == 1:
-            val: Union[bytes, int, str] = args[0]
-            bytesval = parse_bytes(val)
-        elif len(args) > 1:
-            # TODO: each int is 1 byte, check size, create bytesval
-            bytesval = bytes(args)
-
-        if bytesval is None:
-            if cls.length == 0:
-                bytesval = b''
-            else:
-                bytesval = b'\x00' * cls.length
-        if len(bytesval) != cls.length:
-            raise TypeError("BytesN[%d] cannot be initialized with value of %d bytes" % (cls.length, len(bytesval)))
-        return super().__new__(cls, bytesval)
 
-    def serialize(self):
-        from .ssz_impl import serialize
-        return serialize(self, self.__class__)
-
-    def hash_tree_root(self):
-        from .ssz_impl import hash_tree_root
-        return hash_tree_root(self, self.__class__)
-
-
-# SSZ Defaults
-# -----------------------------
-def get_zero_value(typ):
-    if is_uint_type(typ):
-        return 0
-    elif is_list_type(typ):
-        return []
-    elif is_bool_type(typ):
-        return False
-    elif is_vector_type(typ):
-        return typ()
-    elif is_bytesn_type(typ):
-        return typ()
-    elif is_bytes_type(typ):
-        return b''
-    elif is_container_type(typ):
-        return typ(**{f: get_zero_value(t) for f, t in typ.get_fields()})
-    else:
-        raise Exception("Type not supported: {}".format(typ))
-
-
-# Type helpers
-# -----------------------------
-
-
-def infer_type(obj):
-    if is_uint_type(obj.__class__):
-        return obj.__class__
-    elif isinstance(obj, int):
-        return uint64
-    elif isinstance(obj, list):
-        return List[infer_type(obj[0])]
-    elif isinstance(obj, (Vector, Container, bool, BytesN, bytes)):
-        return obj.__class__
-    else:
-        raise Exception("Unknown type for {}".format(obj))
-
-
-def infer_input_type(fn):
```

Comment:
Wait, where do `infer_input_type` and `infer_type` live now?

Reply:
not sure, catching cases where we don't type inputs properly is good imho.
And not sure about wrapping `bool` yet, made it an edge-case for now (see new commits)

Reply:
For now we have:
`TypeWithDefault = Union[DefaultingTypeMeta, bool]`
- parameter type checking in `ParamsMeta` accounts for this, and checks if it's either one of the two options (bool/DefaultingTypeMeta)
- `get_zero_value` checks if the type is a bool, otherwise it calls `typ.default()`.

We could wrap all bools. But unsure if it's worth the type purity (everything having a default, being able to check if something is a "SSZ" type), or if it just bothers us with `Bool(True)`-like problems.

Reply:
got rid of the unnecessary argument, it directly using the value type seems more sensible, and less error prone (you can't have a different type than the value actually is). And it is all nicely checking / type-hinting :)